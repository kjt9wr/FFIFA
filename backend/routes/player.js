const router = require('express').Router();
const Player = require('../models/player.model');

router.route('/').get((req, res) => {
  Player.find()
    .then(player => res.json(player))
    .catch(err => res.status(400).json('Unable to find player: ' + err));
});

router.route('/roster/:ownerID').get((req, res) => {
  Player.find({owner: req.params.ownerID })
    .then(player => res.json(player))
    .catch(err => res.status(400).json('Unable to find player: ' + err));
});

// add player to db
router.route('/add').post((req, res) => {
  const name = req.body.name;
  const price = Number(req.body.price);
  const keep = Boolean(req.body.keep);
  const position = req.body.position;
  const rank = Number(req.body.rank);
  const owner = ownersIDByName[req.body.ownerName];
  const sleeperId = req.body.sleeperId;
  const newPlayer = new Player({ name, price, keep, position, rank, owner, sleeperId });

  newPlayer.save()
  .then(() => res.json(name + ' added! ID: '+ newPlayer._id ))
  .catch(err => res.status(400).json('Unable to add player: ' + err));
});

router.route('/add/sleeperId').post((req, res) => {
  Player.findOne({ name: req.body.name })
    .then(player => {
      player.sleeperId = req.body.sleeperId;
      player.save()
          .then(() => res.json(player.name + ' sleeperId updated!'))
          .catch(err => res.status(400).json('Unable to update player: ' + err));
      })
      .catch(err => res.status(400).json('Unable to find player: ' + err));
}); 


// Updating a keeper
router.route('/update/:pid').post((req, res) => {
  Player.findById(req.params.pid)
    .then(player => {
      player.keep = req.body.keep;
      player.save()
          .then(() => res.json('Player updated!'))
          .catch(err => res.status(400).json('Unable to update player: ' + err));
      })
      .catch(err => res.status(400).json('Unable to find player: ' + err));
  });

  // Get all players for given position, ordered by rank
  router.route('/getAll/:pos').get((req, res) => {
    Player.find({position: req.params.pos })
    .then((data) => { 
      res.send({data: getAllPlayersOrderedByRank(data)}); 
    })
    .catch(err => res.status(400).json('Unable to find player: ' + err));
  })


// Change a player's power ranking
router.route('/rank').put((req, res) => {
  Player.findById(req.body.pid)
    .then(player => {
      player.rank = req.body.rank;
      player.save()
          .then(() => res.json(player.name + ' rank updated!'))
          .catch(err => res.status(400).json('Unable to save player ranking: ' + err));
      })
      .catch(err => res.status(400).json('Unable to find player: ' + err));
  });

  // Set a player's arbitration year
router.route('/arbitration').put((req, res) => {
  Player.findById(req.body.pid)
    .then(player => {
      player.arbYear = req.body.arbYear;
      player.save()
          .then(() => res.json(player.name + ' arbitration year updated!'))
          .catch(err => res.status(400).json('Unable to save player arbitration year: ' + err));
      })
      .catch(err => res.status(400).json('Unable to find player: ' + err));
  });

  // Change a player's keeper class
  router.route('/keeperClass').put((req, res) => {
    Player.findById(req.body.pid)
      .then(player => {
        player.keeperClass = req.body.keeperClass;
        player.save()
            .then(() => res.json(player.name + ' Keeper class updated!'))
            .catch(err => res.status(400).json('Unable to save keeper class : ' + err));
        })
        .catch(err => res.status(400).json('Unable to find player: ' + err));
  });

//reset all keeper classes
router.route('/reset/keeperClass').put((req, res) => {
Player.find().then(player => {
  player.keeperClass = 0; 
  player.save()
  .then(() => res.json(player.name + ' Keeper class updated!'))
  .catch(err => res.status(400).json('Unable to save keeper class : ' + err));
})

  });

// Change a player's supermax status
router.route('/superMax').put((req, res) => {
  Player.findById(req.body.pid)
    .then(player => {
      player.superMax.plan = req.body.superMaxPlan;
      player.superMax.year = req.body.superMaxYear;
      player.keep = true;
      player.save()
          .then(() => res.json(player.name + ' super max status updated!'))
          .catch(err => res.status(400).json('Unable to save player supermax : ' + err));
      })
      .catch(err => res.status(400).json('Unable to find player: ' + err));
  });

// Change a player's owner 
router.route('/update/owner').put((req, res) => {
  const ownerid = ownersIDByName[req.body.ownerName];
    Player.findById(req.body.pid)
    .then(player => {
      player.owner = ownerid;
      player.save()
          .then(() => res.json(player.name + ' owner updated!'))
          .catch(err => res.status(400).json('Unable to update player: ' + err));
      })
      .catch(err => res.status(400).json('Unable to find player: ' + err));
});

// Update roster
router.route('/update/roster/:ownerId').post((req, res) => {
  const playersToUpdate = req.body.players;
  Player.updateMany({
      sleeperId: {
        $in: playersToUpdate
      }}, {
        $set: { owner: req.params.ownerId }
      })
  .then(() => res.json('Updated!'))
});

// Updating a draft result
router.route('/draftResult').put((req, res) => {
  let ownerid = ownersIDByName[req.body.ownerName];
    Player.findById(req.body.pid)
    .then(player => {
      player.owner = ownerid;
      player.price = Math.trunc(req.body.price*1.2);
      player.save()
          .then(() => res.json(player.name + ' drafted by ' + req.body.ownerName +  ' at $' + req.body.price))
          .catch(err => res.status(400).json('Unable to update player: ' + err));
      })
      .catch(err => res.status(400).json('Unable to find player: ' + err));
  });

const getAllPlayersOrderedByRank = (playerList) => {
  return playerList.map((player) => {
    return {
      _id: player.id,
      name: player.name,
      rank: player.rank
   }
  }).sort((a,b) => a.rank - b.rank);
}

const ownersIDByName = {
  "Kevin" : "5e80d724b3bdaf3413316177",
  "Justin" : "5e80d930b3bdaf3413316189",
  "Alex" : "5e80dd6ab3bdaf34133161bd",
  "Luigi" : "5e80da66b3bdaf341331619b",
  "Christian" : "5e80e173b3bdaf3413316213",
  "Matt" : "5e80df96b3bdaf34133161ef",
  "Brent" : "5e80db62b3bdaf34133161ab",
  "Michael" : "5e80de37b3bdaf34133161cf",
  "Nikos" : "5e80dedcb3bdaf34133161dd",
  "Chinmay" : "5e80e07eb3bdaf3413316200",
  "Patrick" : "5e80e1dab3bdaf3413316225",
  "Jeff" : "5e80e1deb3bdaf3413316226",
};

module.exports = router;
