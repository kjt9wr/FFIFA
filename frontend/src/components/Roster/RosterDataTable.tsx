import { Table } from "reactstrap";
import { FranchiseTagDTO, Player } from "../../interfaces/interfaces";
import RosterPlayerRow from "./RosterPlayerRow";

interface RosterDataTableProps {
  roster: Player[];
  franchisePrices: FranchiseTagDTO;
  toggleKeeper: (e: Event) => void;
  editFranchiseMode?: boolean;
  onSubmitFranchise?: (e: Event) => void;
}

/*
 * This component displays the roster table
 */
const RosterDataTable = (props: RosterDataTableProps) => {
  return (
    <Table responsive size="md" hover borderless>
      <thead className="thead-light">
        <tr>
          <th>Position</th>
          <th>Player</th>
          <th>Price</th>
          <th>Keep</th>
          <th>Keep Class</th>
        </tr>
      </thead>
      <tbody>
        {props.roster.map((currentPlayer, index) => {
          return (
            <RosterPlayerRow
              index={index}
              player={currentPlayer}
              key={currentPlayer.sleeperId}
              id={currentPlayer.sleeperId}
              keep={currentPlayer.keep}
              toggleKeeper={props.toggleKeeper}
              franchisePrices={props.franchisePrices}
            />
          );
        })}
      </tbody>
    </Table>
  );
};

export default RosterDataTable;
