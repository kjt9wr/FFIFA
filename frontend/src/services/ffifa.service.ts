import {
  FranchiseTagDTO,
  Owner,
  PenaltyFeeInfo,
  Player,
} from "../interfaces/interfaces";
import { BASE_CAP, getUpcomingSeasonYear } from "../utilities/constants";
import { KEEPER_CLASS_ENUM, POSITION } from "../utilities/enumerations";
import { calculateSuperMaxPrice } from "./supermax.service";

export const determineFinalPriceOfPlayer = (
  player: Player,
  franchisePrices: FranchiseTagDTO
) => {
  switch (player.keeperClass) {
    case 2:
      return getFranchisePrice(player.position, franchisePrices);
    case 3:
      return player.superMax
        ? calculateSuperMaxPrice(
            player.superMax.plan,
            player.superMax.signingYear
          )
        : -99;
    case 4:
      return getFranchisePrice(player.position, franchisePrices);
    default:
      return player.price;
  }
};

const getFranchisePrice = (playerPosition: string, prices: FranchiseTagDTO) => {
  switch (playerPosition) {
    case POSITION.QB:
      return prices.qbFranchisePrice;
    case POSITION.RB:
      return prices.rbFranchisePrice;
    case POSITION.WR:
      return prices.wrFranchisePrice;
    default:
      return prices.teFranchisePrice;
  }
};

export const pickSuperMaxOrKeeperPrice = (player: Player) => {
  return KEEPER_CLASS_ENUM.SUPERMAX === player.keeperClass && player.superMax
    ? calculateSuperMaxPrice(player.superMax.plan, player.superMax.signingYear)
    : player.price;
};

export const increaseKeeperPrice = (draftedPrice: number) => {
  return Math.max(10, Math.trunc(1.2 * draftedPrice));
};

export const calculateLuxaryPotPayout = (penaltyFees: PenaltyFeeInfo[]) => {
  const totalInPot = penaltyFees.reduce(
    (acc, owner) => acc + owner.penaltyFee,
    0
  );

  const nonOffenders = penaltyFees.filter(
    (owner) => owner.penaltyFee === 0
  ).length;

  return Math.trunc(totalInPot / nonOffenders);
};

export const calculateTotalInPot = (penaltyFees: PenaltyFeeInfo[]) => {
  return penaltyFees.reduce(
    (acc: number, owner: PenaltyFeeInfo) => acc + owner.penaltyFee,
    0
  );
};

export const calculateFollowingYearBaseCap = (
  penaltyFees: PenaltyFeeInfo[]
) => {
  const nonoffendersCount = penaltyFees.filter(
    (owner: Owner) => owner.penaltyFee === 0
  ).length;

  const offendersCount = penaltyFees.length - nonoffendersCount;

  return Math.trunc(
    BASE_CAP[getUpcomingSeasonYear()] *
      (1.05 + 0.01 * (nonoffendersCount - offendersCount))
  );
};
