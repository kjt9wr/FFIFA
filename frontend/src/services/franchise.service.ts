import { updatePlayerStatus } from "../api/api.service";
import { Player, UpdatePlayerDTO } from "../interfaces/interfaces";
import * as FFIFAService from "./ffifa.service";

export const get10MostExpensivePerPosition = (
  allKeptPlayers: Player[],
  position: string
) => {
  return allKeptPlayers
    .filter((player: Player) => player.position === position)
    .map((player: Player) => ({
      ...player,
      price: FFIFAService.pickSuperMaxOrKeeperPrice(player),
    }))
    .sort((a: Player, b: Player) => b.price - a.price)
    .slice(0, 10);
};

export const calculateFranchisePrice = (sortedPositionList: Player[]) => {
  const sum = sortedPositionList
    .slice(0, 5)
    .reduce((acc, player) => acc + player.price, 0);
  return Math.trunc(sum / Math.min(sortedPositionList.length, 5));
};

export const removeAllFranchiseTags = async (roster: Player[]) => {
  roster
    .filter((player) => player.keeperClass === 2)
    .forEach(async (player) => {
      const updateDTO: UpdatePlayerDTO = {
        keeperClass: 1,
      };
      await updatePlayerStatus(player.sleeperId, updateDTO);
    });
};

export const getFranchiseEligiblePlayers = (roster: Player[]) => {
  return roster
    .filter((player: Player) => ![3, 4].includes(player.keeperClass))
    .sort(
      (a: Player, b: Player) =>
        a.position.localeCompare(b.position) ||
        b.price - a.price ||
        a.name.localeCompare(b.name)
    );
};
