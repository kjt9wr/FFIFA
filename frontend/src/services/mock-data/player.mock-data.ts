import { Player } from "../../interfaces/interfaces";

export const MAXED_PLAYER: Player = {
  _id: "abc",
  keep: true,
  keeperClass: 3,
  name: "Najee Harris",
  owner: "",
  position: "RB",
  superMax: {
    plan: 5,
    signingYear: 2023,
  },
  price: 10,
  sleeperId: "abc",
};

export const ARBITRATED_TE: Player = {
  _id: "kittle",
  keep: true,
  keeperClass: 4,
  name: "George Kittle",
  owner: "",
  position: "TE",
  price: 50,
  sleeperId: "kittle",
};

// Kept Players

const KEPT_PLAYER = {
  keep: true,
  keeperClass: 1,
  owner: "",
};

export const MOCK_JAYDEN_DANIELS: Player = {
  ...KEPT_PLAYER,
  _id: "jayden_daniels",
  keeperClass: 2,
  name: "Jayden Daniels",
  position: "QB",
  price: 67,
  sleeperId: "jayden_daniels",
};

export const MOCK_LAMAR_JACKSON: Player = {
  ...KEPT_PLAYER,
  _id: "lamar_jackson",
  name: "Lamar Jackson",
  position: "QB",
  price: 37,
  sleeperId: "lamar_jackson",
};

export const MOCK_KYLER_MURRAY: Player = {
  ...KEPT_PLAYER,
  _id: "kyler_murray",
  name: "Kyler Murray",
  position: "QB",
  price: 37,
  sleeperId: "kyler_murray",
};

export const MOCK_PATRICK_MAHOMES: Player = {
  ...KEPT_PLAYER,
  _id: "patrick_mahomes",
  name: "Patrick Mahomes",
  position: "QB",
  price: 37,
  rank: 10,
  sleeperId: "patrick_mahomes",
};

export const MOCK_JALEN_HURTS: Player = {
  ...KEPT_PLAYER,
  _id: "jalen_hurts",
  name: "Jalen Hurts",
  position: "QB",
  price: 37,
  sleeperId: "jalen_hurts",
};

export const MOCK_BIJAN_ROBINSON: Player = {
  ...KEPT_PLAYER,
  _id: "bijan_robison",
  keeperClass: 2,
  name: "Bijan Robinson",
  position: "RB",
  price: 121,
  sleeperId: "bijan_robison",
};

export const MOCK_SAQUON_BARKLEY: Player = {
  ...KEPT_PLAYER,
  _id: "saquon_barkley",
  keeperClass: 2,
  name: "Saquon Barkley",
  position: "RB",
  price: 100,
  sleeperId: "saquon_barkley",
};

export const MOCK_JAHMYR_GIBBS: Player = {
  ...KEPT_PLAYER,
  _id: "jahmyr_gibbs",
  keeperClass: 2,
  name: "Jahmyr Gibbs",
  position: "RB",
  price: 98,
  sleeperId: "jahmyr_gibbs",
};

export const MOCK_JOSH_JACOBS: Player = {
  ...KEPT_PLAYER,
  ...KEPT_PLAYER,
  _id: "josh_jacobs",
  name: "Josh Jacobs",
  position: "RB",
  price: 81,
  rank: 12,
  sleeperId: "josh_jacobs",
};

export const MOCK_DERRICK_HENRY: Player = {
  ...KEPT_PLAYER,
  _id: "derrick_henry",
  name: "Derrick Henry",
  position: "RB",
  price: 79,
  firstKeepYear: 2023,
  rank: 6,
  sleeperId: "derrick_henry",
};

export const MOCK_NAJEE_HARRIS: Player = {
  ...KEPT_PLAYER,
  _id: "najee_harris",
  keeperClass: 3,
  name: "Najee Harris",
  sleeperId: "najee_harris",
  position: "RB",
  owner: "Patrick",
  superMax: {
    plan: 5,
    signingYear: 2022,
  },
  price: 130,
};

export const MOCK_BROCK_BOWERS: Player = {
  ...KEPT_PLAYER,
  _id: "brock_bowers",
  keeperClass: 2,
  name: "Brock Bowers",
  position: "TE",
  price: 37,
  sleeperId: "brock_bowers",
};

export const MOCK_SAM_LAPORTA: Player = {
  ...KEPT_PLAYER,
  _id: "sam_laporta",
  name: "Sam Laporta",
  position: "TE",
  price: 12,
  rank: 4,
  sleeperId: "sam_laporta",
};

export const MOCK_DRAKE_LONDON: Player = {
  ...KEPT_PLAYER,
  _id: "drake_london",
  name: "Drake London",
  position: "WR",
  price: 64,
  keeperClass: 2,
  sleeperId: "drake_london",
};

export const MOCK_AJ_BROWN: Player = {
  ...KEPT_PLAYER,
  _id: "aj_brown",
  name: "AJ Brown",
  position: "WR",
  price: 61,
  keeperClass: 2,
  sleeperId: "aj_brown",
};

export const MOCK_CEEDEE_LAMB: Player = {
  ...KEPT_PLAYER,
  _id: "ceedee_lamb",
  name: "CeeDee Lamb",
  position: "WR",
  price: 56,
  sleeperId: "ceedee_lamb",
};

export const MOCK_TYREEK_HILL: Player = {
  ...KEPT_PLAYER,
  _id: "tyreek_hill",
  name: "Tyreek Hill",
  position: "WR",
  price: 56,
  rank: 10,
  sleeperId: "tyreek_hill",
};

export const MOCK_JAMARR_CHASE: Player = {
  ...KEPT_PLAYER,
  _id: "jamarr_chase",
  name: "Jamarr Chase",
  position: "WR",
  price: 52,
  firstKeepYear: 2022,
  sleeperId: "jamarr_chase",
};

export const MOCK_MALIK_NABERS: Player = {
  ...KEPT_PLAYER,
  _id: "malik_nabers",
  name: "Malik Nabers",
  position: "WR",
  price: 51,
  sleeperId: "malik_nabers",
};

export const MOCK_JALEN_COKER: Player = {
  ...KEPT_PLAYER,
  _id: "jalen_coker",
  name: "Jalen Coker",
  position: "WR",
  price: 10,
  rank: 30,
  sleeperId: "jalen_coker",
};

export const MOCK_CHRIS_OLAVE: Player = {
  _id: "chris_olave",
  keep: false,
  keeperClass: 1,
  name: "Chris Olave",
  owner: "",
  position: "WR",
  price: 62,
  sleeperId: "chris_olave",
};

export const MOCKED_ROSTER = [
  MOCK_AJ_BROWN,
  MOCK_CEEDEE_LAMB,
  MOCK_JALEN_HURTS,
  MOCK_CHRIS_OLAVE,
];

export const MOCKED_ROSTER_WITH_SUPERMAX = [
  ...MOCKED_ROSTER,
  MOCK_NAJEE_HARRIS,
];
