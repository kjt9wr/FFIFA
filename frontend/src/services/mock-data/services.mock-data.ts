import { FreeAgentStorage, Player } from "../../interfaces/interfaces";
import {
  MAXED_PLAYER,
  MOCK_AJ_BROWN,
  MOCK_BIJAN_ROBINSON,
  MOCK_BROCK_BOWERS,
  MOCK_CEEDEE_LAMB,
  MOCK_DERRICK_HENRY,
  MOCK_DRAKE_LONDON,
  MOCK_JAHMYR_GIBBS,
  MOCK_JALEN_COKER,
  MOCK_JALEN_HURTS,
  MOCK_JAMARR_CHASE,
  MOCK_JAYDEN_DANIELS,
  MOCK_JOSH_JACOBS,
  MOCK_KYLER_MURRAY,
  MOCK_LAMAR_JACKSON,
  MOCK_MALIK_NABERS,
  MOCK_NAJEE_HARRIS,
  MOCK_PATRICK_MAHOMES,
  MOCK_SAM_LAPORTA,
  MOCK_SAQUON_BARKLEY,
  MOCK_TYREEK_HILL,
} from "./player.mock-data";

export const FRANCHISE_TAG_SAMPLE_DATA = {
  qbFranchisePrice: 25,
  keptQBs: [],
  rbFranchisePrice: 75,
  keptRBs: [MOCK_DERRICK_HENRY, MAXED_PLAYER],
  wrFranchisePrice: 50,
  keptWRs: [],
  teFranchisePrice: 30,
  keptTEs: [],
};

export const MOCKED_PENALTY_FEES = [
  { name: "Kevin", penaltyFee: 20 },
  { name: "Justin", penaltyFee: 50 },
  { name: "Matt", penaltyFee: 0 },
  { name: "Luigi", penaltyFee: 0 },
  { name: "Alex", penaltyFee: 30 },
];

export const KEPT_PLAYERS: Player[] = [
  MOCK_JAYDEN_DANIELS,
  MOCK_JALEN_HURTS,
  MOCK_KYLER_MURRAY,
  MOCK_LAMAR_JACKSON,
  MOCK_PATRICK_MAHOMES,
  MOCK_BIJAN_ROBINSON,
  MOCK_SAQUON_BARKLEY,
  MOCK_JAHMYR_GIBBS,
  MOCK_JOSH_JACOBS,
  MOCK_DERRICK_HENRY,
  MOCK_NAJEE_HARRIS,
  MOCK_BROCK_BOWERS,
  MOCK_SAM_LAPORTA,
  MOCK_AJ_BROWN,
  MOCK_DRAKE_LONDON,
  MOCK_CEEDEE_LAMB,
  MOCK_TYREEK_HILL,
  MOCK_JAMARR_CHASE,
  MOCK_MALIK_NABERS,
  MOCK_JALEN_COKER,
];

export const MOCK_FREE_AGENTS: FreeAgentStorage = {
  availableQBs: [MOCK_PATRICK_MAHOMES],
  availableRBs: [MOCK_DERRICK_HENRY, MOCK_JOSH_JACOBS],
  availableWRs: [MOCK_TYREEK_HILL, MOCK_JALEN_COKER],
  availableTEs: [MOCK_SAM_LAPORTA],
};
