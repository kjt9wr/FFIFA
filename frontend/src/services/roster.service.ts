import { fetchSingleOwner } from "../api/api.service";
import { Player, FranchiseTagDTO } from "../interfaces/interfaces";
import { getUpcomingYearIndex } from "../utilities/constants";
import { determineFinalPriceOfPlayer } from "./ffifa.service";

export const calculateLuxaryTaxLine = (cap: number) => Math.trunc(cap * 0.55);

export const calculatePenaltyFee = (
  roster: Player[],
  franchisePrices: FranchiseTagDTO,
  maxCap: number
) => {
  const penaltyFee =
    calculateTotalKeeperPrice(roster, franchisePrices) -
    calculateLuxaryTaxLine(maxCap);
  return penaltyFee > 0 ? penaltyFee : 0;
};

export const calculateTotalKeeperPrice = (
  roster: Player[],
  franchisePrices: FranchiseTagDTO
) => {
  return roster
    .filter((keptPlayer) => keptPlayer.keep)
    .reduce(
      (acc, player) =>
        acc + determineFinalPriceOfPlayer(player, franchisePrices),
      0
    );
};

export const getOwnersCap = async (name: string) => {
  const owner = await fetchSingleOwner(name);
  console.log(owner.data);
  return owner.data[0].cap[getUpcomingYearIndex()];
};
