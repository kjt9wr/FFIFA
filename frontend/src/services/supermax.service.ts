import { BASE_CAP, getUpcomingSeasonYear } from "../utilities/constants";

const THREE_YEAR_RATES = [0.25, 0.2375, 0.2256];
const FOUR_YEAR_RATES = [0.2375, 0.2256, 0.2143, 2036];
const FIVE_YEAR_RATES = [0.225, 0.2138, 0.2031, 0.1929, 0.1833];
const upcomingYear = getUpcomingSeasonYear();
export const getCurrentSuperMaxYear = (signingYear: number) => {
  return Number(upcomingYear) - signingYear;
};

export const calculateSuperMaxPrice = (
  durationInYears: number,
  signingYear: number
) => {
  const currentYear = getCurrentSuperMaxYear(signingYear);

  switch (durationInYears) {
    case 3:
      return Math.trunc(
        BASE_CAP[upcomingYear] * THREE_YEAR_RATES[currentYear - 1]
      );
    case 4:
      return Math.trunc(
        BASE_CAP[upcomingYear] * FOUR_YEAR_RATES[currentYear - 1]
      );
    case 5:
      return Math.trunc(
        BASE_CAP[upcomingYear] * FIVE_YEAR_RATES[currentYear - 1]
      );
    default:
      return 0;
  }
};
