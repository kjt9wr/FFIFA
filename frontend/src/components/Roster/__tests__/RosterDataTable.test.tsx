import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { MOCKED_ROSTER_WITH_SUPERMAX } from "../../../services/mock-data/player.mock-data";
import { FRANCHISE_TAG_SAMPLE_DATA } from "../../../services/mock-data/services.mock-data";
import RosterDataTable from "../RosterDataTable";

const mockToggleKeeper = jest.fn();
describe("roster data table", () => {
  it("renders successfully", () => {
    const view = render(
      <RosterDataTable
        roster={MOCKED_ROSTER_WITH_SUPERMAX}
        franchisePrices={FRANCHISE_TAG_SAMPLE_DATA}
        toggleKeeper={mockToggleKeeper}
      />
    );

    expect(view).toMatchSnapshot();
  });

  it("keeps player on click", async () => {
    render(
      <RosterDataTable
        roster={MOCKED_ROSTER_WITH_SUPERMAX}
        franchisePrices={FRANCHISE_TAG_SAMPLE_DATA}
        toggleKeeper={mockToggleKeeper}
      />
    );

    userEvent.click(await screen.findByTitle("AJ Brown"));

    await waitFor(() => {
      expect(mockToggleKeeper).toHaveBeenCalled();
    });
  });
});
