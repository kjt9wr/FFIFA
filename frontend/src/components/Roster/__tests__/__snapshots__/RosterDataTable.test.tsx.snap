// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`roster data table renders successfully 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="table-responsive"
      >
        <table
          class="table table-md table-borderless table-hover"
        >
          <thead
            class="thead-light"
          >
            <tr>
              <th>
                Position
              </th>
              <th>
                Player
              </th>
              <th>
                Price
              </th>
              <th>
                Keep
              </th>
              <th>
                Keep Class
              </th>
            </tr>
          </thead>
          <tbody>
            <tr
              class="player-table-odd"
            >
              <td>
                WR
              </td>
              <td>
                 
                AJ Brown
                 
              </td>
              <td>
                <span
                  class="previousValue"
                >
                   
                  61
                </span>
                 
                50
              </td>
              <td>
                <input
                  checked=""
                  class="form-check-input"
                  id="aj_brown"
                  title="AJ Brown"
                  type="checkbox"
                />
              </td>
              <td>
                Franchise Tag
              </td>
            </tr>
            <tr
              class="player-table-even"
            >
              <td>
                WR
              </td>
              <td>
                 
                CeeDee Lamb
                 
              </td>
              <td>
                 
                56
              </td>
              <td>
                <input
                  checked=""
                  class="form-check-input"
                  id="ceedee_lamb"
                  title="CeeDee Lamb"
                  type="checkbox"
                />
              </td>
              <td />
            </tr>
            <tr
              class="player-table-odd"
            >
              <td>
                QB
              </td>
              <td>
                 
                Jalen Hurts
                 
              </td>
              <td>
                 
                37
              </td>
              <td>
                <input
                  checked=""
                  class="form-check-input"
                  id="jalen_hurts"
                  title="Jalen Hurts"
                  type="checkbox"
                />
              </td>
              <td />
            </tr>
            <tr
              class="player-table-even"
            >
              <td>
                WR
              </td>
              <td>
                 
                Chris Olave
                 
              </td>
              <td>
                 
                62
              </td>
              <td>
                <input
                  class="form-check-input"
                  id="chris_olave"
                  title="Chris Olave"
                  type="checkbox"
                />
              </td>
              <td />
            </tr>
            <tr
              class="player-table-odd"
            >
              <td>
                RB
              </td>
              <td>
                 
                Najee Harris
                 
              </td>
              <td>
                 
                76
              </td>
              <td>
                <input
                  checked=""
                  class="form-check-input"
                  disabled=""
                  id="najee_harris"
                  title="Najee Harris"
                  type="checkbox"
                />
              </td>
              <td>
                Supermax
                <span>
                   
                  Y
                  3
                  /
                  5
                </span>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="table-responsive"
    >
      <table
        class="table table-md table-borderless table-hover"
      >
        <thead
          class="thead-light"
        >
          <tr>
            <th>
              Position
            </th>
            <th>
              Player
            </th>
            <th>
              Price
            </th>
            <th>
              Keep
            </th>
            <th>
              Keep Class
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class="player-table-odd"
          >
            <td>
              WR
            </td>
            <td>
               
              AJ Brown
               
            </td>
            <td>
              <span
                class="previousValue"
              >
                 
                61
              </span>
               
              50
            </td>
            <td>
              <input
                checked=""
                class="form-check-input"
                id="aj_brown"
                title="AJ Brown"
                type="checkbox"
              />
            </td>
            <td>
              Franchise Tag
            </td>
          </tr>
          <tr
            class="player-table-even"
          >
            <td>
              WR
            </td>
            <td>
               
              CeeDee Lamb
               
            </td>
            <td>
               
              56
            </td>
            <td>
              <input
                checked=""
                class="form-check-input"
                id="ceedee_lamb"
                title="CeeDee Lamb"
                type="checkbox"
              />
            </td>
            <td />
          </tr>
          <tr
            class="player-table-odd"
          >
            <td>
              QB
            </td>
            <td>
               
              Jalen Hurts
               
            </td>
            <td>
               
              37
            </td>
            <td>
              <input
                checked=""
                class="form-check-input"
                id="jalen_hurts"
                title="Jalen Hurts"
                type="checkbox"
              />
            </td>
            <td />
          </tr>
          <tr
            class="player-table-even"
          >
            <td>
              WR
            </td>
            <td>
               
              Chris Olave
               
            </td>
            <td>
               
              62
            </td>
            <td>
              <input
                class="form-check-input"
                id="chris_olave"
                title="Chris Olave"
                type="checkbox"
              />
            </td>
            <td />
          </tr>
          <tr
            class="player-table-odd"
          >
            <td>
              RB
            </td>
            <td>
               
              Najee Harris
               
            </td>
            <td>
               
              76
            </td>
            <td>
              <input
                checked=""
                class="form-check-input"
                disabled=""
                id="najee_harris"
                title="Najee Harris"
                type="checkbox"
              />
            </td>
            <td>
              Supermax
              <span>
                 
                Y
                3
                /
                5
              </span>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
